

rule htseq_count:
    conda:
        "../envs/htseq.yml"
    input:
        "reference/reference_genome.gff",
        "samples/{sample}/mapping/bwa/reference_genome.bam"
    output:
        "samples/{sample}/rnaseq/htseq/counts.txt"
    shell:
        """
        htseq-count --stranded=no -t CDS --idattr=locus_tag -f bam {input[1]} {input[0]} > {output[0]}
        """

rule htseq_count_subsets:
    conda:
        "../envs/htseq.yml"
    input:
        "reference/reference_genome.gff",
        "samples/{sample}/mapping/bwa/sample_{count}/reference_genome.bam"
    output:
        "samples/{sample}/rnaseq/htseq/sample_{count}_counts.txt"
    shell:
        """
        htseq-count --stranded=no -t CDS --idattr=locus_tag -f bam {input[1]} {input[0]} > {output[0]}
        """

rule counts_summary:
    conda:
        "../envs/htseq.yml"
    input:
        htseq_files = expand("samples/{sample}/rnaseq/htseq/counts.txt", sample=read_naming.keys())
    output:
        "report/rnaseq/htseq/summary_all.tab"
    script: "scripts/merge_htseq.py"
