

rule assess_mapping_with_qualimap:
    conda:
        "../envs/qualimap.yml"
    input:
        # samples/{sample}/mapping/bwa/{ref}.bam
        bam = "samples/{sample}/mapping/bwa/reference_genome.bam",
        bam_bai = "samples/{sample}/mapping/bwa/reference_genome.bam.bai"
    output:
        report = "report/qualimap/{sample}/bwa/reference_genome/qualimapReport.html",
        genome_results = "report/qualimap/{sample}/bwa/reference_genome/genome_results.txt",
    log:
        logging_folder+"samples/{sample}/logs/bwa/reference_genome/qualimap.txt",
    shell:
        """
        qualimap bamqc -bam {input[bam]} -outdir $(dirname {output}) > {log[0]}
        if [ ! -f {output[report]} ]
        then
            touch {output[report]}
            touch {output[genome_results]}
        fi
        """

rule mapping_stats:
    input:
        expand("report/qualimap/{sample}/bwa/reference_genome/genome_results.txt", sample=read_naming.keys()),
    output:
        "report/mapping_summary.tsv"
    shell:
        """
        for file in {input}; do
            sample=`grep "outfile" $file  | sed 's+.*qualimap/++' | sed 's+/.*++'`
            n_reads=`grep "number of reads" $file  | sed 's/.*= //' | sed 's/,//g'`
            n_mapped_reads=`grep "number of mapped reads" $file  | sed 's/.*= //' | sed 's/,//g' | sed 's/ .*//'`
            percent_mappd=`grep "number of mapped reads" $file  | sed 's/.*= //' | sed 's/,//g' | sed 's/.* //' | sed 's/.* //' | sed 's/(//' | sed 's/)//' `
            echo -e "$sample\t$n_reads\t$n_mapped_reads\t$percent_mappd" >> report/mapping_summary.tsv
        """